feat: handle enrollments for "Invite Only" courses

tecoholic opened the pull request:
### Description

This PR introduces some improvements to the 'Manage Learners' page of Enterprise Customer admin that will allow enrolling learners into "Invite Only" courses.

Refer #1745 for full details about the issue.

### Testing instructions

* Follow the testing instructions in #1745 and reproduce the issue
* Switch to the PR branch
* Try enrolling learners into the course marked as "Invite Only", with the *Force Enrollment* checkbox checked, this time it should work without throwing an error.
* Test with both a learner who already has a registered account and an email which is not registered in the system to ensure, "enrollment on registration" also works for "Invite Only" courses.

![image](https://github.com/openedx/edx-enterprise/assets/383862/e6e17c83-6c69-459c-b8e8-9e22756562f3)


**Merge checklist:**
- [ ] Any new requirements are in the right place (do **not** manually modify the `requirements/*.txt` files) - **No new requirements added**
    - `base.in` if needed in production but edx-platform doesn't install it
    - `test-master.in` if edx-platform pins it, with a matching version
    - `make upgrade && make requirements` have been run to regenerate requirements
- [ ] `make static` has been run to update webpack bundling if any static content was updated - **Running make static didn't introduce any changes**
- [ ] `./manage.py makemigrations` has been run - **No DB migrations introduced**
    - Checkout the [Database Migration](https://openedx.atlassian.net/wiki/spaces/AC/pages/23003228/Everything+About+Database+Migrations) Confluence page for helpful tips on creating migrations.
    - *Note*: This **must** be run if you modified any models.
      - It may or may not make a migration depending on exactly what you modified, but it should still be run.
    - This should be run from either a venv with all the lms/edx-enterprise requirements installed or if you checked out edx-enterprise into the src directory used by lms, you can run this command through an lms shell.
        - It would be `./manage.py lms makemigrations` in the shell.
- [ ] [Version](https://github.com/openedx/edx-enterprise/blob/master/enterprise/__init__.py) bumped
- [ ] [Changelog](https://github.com/openedx/edx-enterprise/blob/master/CHANGELOG.rst) record added
- [ ] Translations updated (see docs/internationalization.rst but also this isn't blocking for merge atm)

**Post merge:**
- [ ] Tag pushed and a new [version](https://github.com/openedx/edx-enterprise/releases) released
    - *Note*: Assets will be added automatically. You just need to provide a tag (should match your version number) and title and description.
- [x] After versioned build finishes in [GitHub Actions](https://github.com/openedx/edx-enterprise/actions), verify version has been pushed to [PyPI](https://pypi.org/project/edx-enterprise/)
    - Each step in the release build has a condition flag that checks if the rest of the steps are done and if so will deploy to PyPi.
    (so basically once your build finishes, after maybe a minute you should see the new version in PyPi automatically (on refresh))
- [ ] PR created in [edx-platform](https://github.com/openedx/edx-platform) to upgrade dependencies (including edx-enterprise)
    - This **must** be done after the version is visible in PyPi as `make upgrade` in edx-platform will look for the latest version in PyPi.
    - Note: the edx-enterprise constraint in edx-platform **must** also be bumped to the latest version in PyPi.

openedx-webhooks commented:
Thanks for the pull request, @tecoholic!  Please note that it may take us up to several weeks or months to complete a review and merge your PR.

Feel free to add as much of the following information to the ticket as you can:
- supporting documentation
- [Open edX discussion forum threads](https://discuss.openedx.org/)
- timeline information ("this must be merged by XX date", and why that is)
- partner information ("this is a course on edx.org")
- any other information that can help Product understand the context for the PR

All technical communication about the code itself will be done via the GitHub pull request interface. As a reminder, [our process documentation is here](http://edx-developer-guide.readthedocs.org/en/latest/process/overview.html).


Please let us know once your PR is ready for our review and all tests are green.




<!-- comment:external_pr -->
<!-- jenkins ok to test -->
<!-- data: eyJkcmFmdCI6IGZhbHNlfQ== -->

tecoholic commented on enterprise/static/enterprise/js/manage_learners.js:
```diff
@@ -134,11 +140,26 @@ function loadPage() {
+                "this should be enabled for the enrollments to go through in those courses."
+            );
+        }
+    });
```
This part is has a potential side effect of getting learners enrolled into courses which might be closed for reasons other than being marked "Invite Only". Suggestions for possible solutions for this side-effect are welcome.

Agrendalath replied:
@tecoholic, this sounds like a reasonable tradeoff.

A proper solution I can think of would be adding more checks around [this code](https://github.com/openedx/edx-platform/blob/065ba5fd7360b20756dce81be0449976a0fb596e/common/djangoapps/student/models/course_enrollment.py#L690-L696) to verify if the requester is `staff` (e.g., with `crum.get_current_user`) and if the course is `invitation_only`.

Currently, the [EnrollmentListView](https://github.com/openedx/edx-platform/blob/065ba5fd7360b20756dce81be0449976a0fb596e/openedx/core/djangoapps/enrollments/views.py#L446-L446) does not support invitation only courses. This problem does not happen in the Instructor Dashboard because it doesn't invoke access checks [when calling the `enroll` method](https://github.com/openedx/edx-platform/blob/065ba5fd7360b20756dce81be0449976a0fb596e/common/djangoapps/student/models/course_enrollment.py#L772-L772). Modifying the `enroll`  method is a high-risk change, and it seems that the API will need some changes in the near future anyway - when the Instructor Dashboard is converted into an MFE (it could be an excellent opportunity to eliminate many legacy behaviors there, too). Therefore, the solution from this PR is "good enough" at the moment.

Please add more context as a comment in the code to explain why we are introducing this functionality. Feel free to use this explanation if you'd like.

tecoholic replied:
Thanks for the detailed explanation. I tried to work as much as possible using the existing platform APIs and didn't want to introduce any changes on the platform, as I wasn't sure about the wider implications it might have. I am glad you found it a reasonable tradeoff as well. 

I have added your explanation as a part of the comment in the code. :)

e0d commented:
@tecoholic it looks like the branch has conflicts, can you have a look?

tecoholic commented:
@e0d I have rebased the PR branch and bumped the version in `enterprise/__init__.py`. Conflicts are now resolved.

e0d commented:
@tecoholic Looks like there are some conflicts that need to be resolved, can you have a look?

tecoholic commented:
@e0d I have rebased it on master and fixed the conflicts. Can you kindly approve the tests to run?

e0d commented:
@tecoholic Tests approved.

tecoholic commented:
@e0d Looks like all checks have passed. Can you kindly see if this can be reviewed and merged?

mphilbrick211 commented:
Hi @macdiesel and @openedx/enterprise-titans! Could someone please review and merge this if all looks good? Thanks!

tecoholic commented:
@mphilbrick211 Hi, can you kindly get someone to take a look at this?

mphilbrick211 commented:
Hi @macdiesel and @openedx/enterprise-titans! Friendly ping on this :)

In the meantime, @tecoholic, would you mind resolving the branch conflicts? Thanks!

tecoholic commented:
@mphilbrick211 Done. All good w.r.t conflicts

mphilbrick211 commented:
Hi @tecoholic - tests are running again, but there are some more branch conflicts that have popped up.

tecoholic commented:
@mphilbrick211 I think the conflicts will continue to happen on the changelog and the version number as long as things are changing on the master and it would be a daily task to keep resolving them. So, I am think we could do either of the 2 things:
1. Kindly ping me when someone picks this up for review and I can resolve the conflicts
2. Or the reviewer can resolve it as long as it is just the change-log entry and the version number and it doesn't touch any code.

What do you think?

tecoholic commented:
> @tecoholic I've approved the tests, but looks like there's a conflict.

@e0d I have fixed the conflicts.

e0d commented:
@tecoholic I've run the tests, but there are conflicts that require resolution.  For easy merges, we typically do them from the GitHub UI, but Demid pointed out this doesn't work with organizational forks, https://github.com/orgs/community/discussions/5634

tecoholic commented:
@e0d Hi, I have resolved the conflict. If it runs into conflict again, I guess it wouldn't be possible for you to make changes as Demid pointed out. Would like me to squash the changed into a single commit, that would allow you to cherry-pick and resolve conflicts locally and merge?

e0d commented:
@tecoholic I don't think I'm going to be able to take over conflict management.

I did notice that a number of the OpenCraft PRs are broken because of a doc formatting issue.  See my commit [here](https://github.com/openedx/edx-enterprise/pull/1790).

tecoholic commented:
@e0d Hi, since the conflicts have been occurring on the version bump and the changelog, I have removed them from the PR and rebased it to the latest master. Kindly see if this helps.

Agrendalath commented:
@tecoholic, since the upstream review for this didn't start yet, you can squash the commits. It will be easier to cherry-pick this to our Palm branch.

e0d commented:
@mphilbrick211 this branch will need a rebase when the reviewers are ready to look at it, but the tests are all green.  Removing the  label.

mphilbrick211 commented:
Hi @openedx/enterprise-titans! Following up to see if we could please get this reviewed/merged. Thanks!

Agrendalath commented:
@mphilbrick211, I believe it would make sense to move forward with [this document](https://docs.google.com/document/d/1zQAWJbrmrZsCYGpfIa_mOB_ZDREFjECeyhEoUlzO_fw) first. Without a timeline for these reviews, the branch conflicts will appear regularly.

cc: @tecoholic

mphilbrick211 commented:
> @mphilbrick211, I believe it would make sense to move forward with [this document](https://docs.google.com/document/d/1zQAWJbrmrZsCYGpfIa_mOB_ZDREFjECeyhEoUlzO_fw) first. Without a timeline for these reviews, the branch conflicts will appear regularly.
> 
> cc: @tecoholic

Sounds good - thanks, @Agrendalath!

mphilbrick211 commented:
Hi @openedx/enterprise-titans! Is there an ETA on when someone could please review this? Thanks!

mphilbrick211 commented:
> Hi @openedx/enterprise-titans! Is there an ETA on when someone could please review this? Thanks!

Hi @openedx/enterprise-titans - following up on this!

tecoholic commented:
@openedx/enterprise-titans The PR has been rebased on master and all the checks are passing. Kindly review.

cc: @e0d

mphilbrick211 commented:
Hi @openedx/2u-titans! Are you still reviewing pull requests? If so, could you please review / merge this for us? Thanks!

mphilbrick211 commented:
Closing due to inactivity.
